server/src/main/java/org/opensearch/search/backpressure/TaskCancellation.java

```diff
@@ -0,0 +1,100 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.search.backpressure;
+
+import org.opensearch.search.backpressure.trackers.ResourceUsageTracker;
+import org.opensearch.tasks.CancellableTask;
+
+import java.util.List;
+import java.util.function.LongSupplier;
+import java.util.stream.Collectors;
+
+/**
+ * TaskCancellation is a wrapper for a task and its cancellation reasons.
+ */
+public class TaskCancellation implements Comparable<TaskCancellation> {
+    private final CancellableTask task;
+    private final List<Reason> reasons;
+    private final LongSupplier timeNanosSupplier;
+
+    public TaskCancellation(CancellableTask task, List<Reason> reasons, LongSupplier timeNanosSupplier) {
+        this.task = task;
+        this.reasons = reasons;
+        this.timeNanosSupplier = timeNanosSupplier;
+    }
+
+    public CancellableTask getTask() {
+        return task;
+    }
+
+    public List<Reason> getReasons() {
+        return reasons;
+    }
+
+    public String getReasonString() {
+        return reasons.stream().map(Reason::getMessage).collect(Collectors.joining(", "));
+    }
+
+    /**
+     * Cancels the task and increments the cancellation counters for all breaching resource usage trackers.
+     */
+    public void cancel() {
+        task.cancel("resource consumption exceeded [" + getReasonString() + "]");
+        reasons.forEach(reason -> reason.getTracker().incrementCancellations());
+    }
```

Comment:
1. should this also cancel child tasks?
2. Can this be utilized outside the context of the current search cancellation use case? If yes should we move this out to commons

Reply:
1. The `TaskCancellationService` shall take care of cancelling child tasks. This would happen for tasks that have `shouldCancelChildrenOnCancellation()` overridden to return `true`.
2. That's a fair point; task cancellation can work with any arbitrary (cancellable) tasks.